Introducing the HostService

csillag opened the pull request:
Currently, the show/hide frame functionality is not directly
available for the sidebar code.

However, internally, then annotation-ui-sync service implements
the necessary features.

This commit exposes those implementations as methods on the AnnotationUI
service, which is the top-level entry point for all similar APIs for
controlling the state of the UI.

   * * *

The changes are minimal. 2 lines of code. 
This will make the direct linking code cleaner.

tilgovi commented:
The annotation UI service should not know about frames. Where are you needing this that you can't access the AnnotationUISync service directly?

My preferred approach would be to extract the hide and show into a different service that formalizes the widget functions. Right now, that's blended with annotation UI state, which for all anyone knows could be synchronized with frames that are not nested one within another.

I think there's a separation that's missing and this PR just leaks that problem further away from where it should be isolated.

csillag commented:
> Where are you needing this that you can't access the AnnotationUISync service directly?

My problem is the the `AnnotationUISync` service is not even accessibly publicly anywhere. It just does it's think in the background, but it's never publicly declared or made available to the other pieces of the system.

csillag commented:
> It just does it's think in the background, 

@csillag sighs. Thin**g**.

tilgovi commented:
It's registered with Angular's dependency injector. What's the problem?

csillag commented:
> It's registered with Angular's dependency injector. What's the problem?

We fire it up an instance [here](https://github.com/hypothesis/h/blob/ad189736739a7bab793198f179539b99ae535002/h/static/scripts/cross-frame-service.coffee#L70), but then we don't do anything with the variable. Therefore, it's not accessible.

Furthermore, the AnnotationUISync service offers zero public APIs. Hence, I figured we don't want to export APIs there, and access those directly.

Do we?

tilgovi commented:
Right. My point is that you don't need to export it on a different object.
If it's worthy of export, you can do it on the service in which it's
defined.

You have competing impulses and you're not listening well enough to see
what they tell you together:

- you want show/hide exposed
- you want to preserve the background status of the sync service

The answer is that the widget state is not a background, implicit state. If
it belonged in the AnnotationUI you would have moved it there. It doesn't,
though. Perhaps it belongs in its own place, explicitly dealing with the
widget host communication.
On Mar 13, 2015 2:49 PM, "Kristof Csillag" <notifications@github.com> wrote:

> It's registered with Angular's dependency injector. What's the problem?
>
> We fire it up an instance here
> <https://github.com/hypothesis/h/blob/ad189736739a7bab793198f179539b99ae535002/h/static/scripts/cross-frame-service.coffee#L70>,
> but then we don't do anything with the variable. Therefore, it's not
> accessible.
>
> Furthermore, the AnnotationUISync service offers zero public APIs. Hence,
> I figured we don't want to export APIs there, and access those directly.
>
> Do we?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hypothesis/h/pull/2038#issuecomment-79449470>.
>

csillag commented:
> My preferred approach would be to extract the hide and show into a different service 

Maybe like this? (See current code.)

The thing I am not sure about is how to properly inject the bridge instance, since both the HostService and the bridge are supposed to be singletons, but the bridge itself is not available as a service, only the bridge factory ... hence the current workaround, where I "inject" the bridge back to the host service at the cross-frame service. Probably not the best way to do that.

   * * *

Yes, the tests are broken; I'll fix them (again) when we finished discussing the right distribution of responsibilities between the various services, controllers, etc.

tilgovi commented:
The only thing that instantiates `Bridge` is the cross frame service. I don't see any reason why it can't receive a bridge instance and then the bridge itself can be constructed by an angular factory.

tilgovi commented:
The host service should be injecting the bridge and adding its callbacks. The annotation ui sync should not be injecting the host. The whole purpose of this exercise was to take the host concerns _out_ of annotation ui sync, not just wrap them in another object.

csillag commented:
> The host service should be injecting the bridge and adding its callbacks.

Oh, I see your point now. One question: if we do this, the host service and the AnnotationUISync service will both want to add a callback to the `showAnnotations` event. (One wants to show the sidebar; the other wants to select the correct annotations.) However, the bridge currently supports only one method binding per event. ("Listener <whatever> already bound in Bridge.")

What approach would you suggest around this? Should we extend the bridge's callback method binding support to a full-blown event listener system, or should we do something else?

csillag commented:
> The only thing that instantiates `Bridge` is the cross frame service. I don't see any reason why it can't receive a bridge instance and then the bridge itself can be constructed by an angular factory.

Maybe because constructing the bridge instance needs some runtime parameters, and whoever designed the current implementation, he felt that those parameters (like `scope: 'annotator:bridge'`) don't belong to an angular factory?

tilgovi commented:
Or have the guest send two messages.

tilgovi commented:
The runtime parameters are constant.

csillag commented:
> The host service should be injecting the bridge and adding its callbacks. The annotation ui sync should not be injecting the host. The whole purpose of this exercise was to take the host concerns out of annotation ui sync, [...]

OK, AnnotationUISync and HostService have been completely separated now.
(Along the way, I have also eliminated a useless cross-frame message - showEditor)

csillag commented:
Renamed PR to reflect on changing scope.

tilgovi commented on h/static/scripts/directives/markdown.coffee:
```diff
@@ -336,7 +336,7 @@ markdown = ['$filter', '$sanitize', '$sce', '$timeout', ($filter, $sanitize, $sc
     inputEl.bind 'blur change keyup', ->
       ctrl.$setViewValue inputEl.val()
-      scope.$digest()
+#      scope.$digest()  # TODO: this shouldn't be a problem
```
Not in this PR, please. Also, I think this is probably right since we're using `.bind` on the element directly.

csillag replied:
>> \# scope.$digest() 

> Not in this PR, please.

I had no intention to leave it this way by merge time; I had to temporarly comment this out, because I was getting "already in $digest" errors.

csillag replied:
> I had to temporarly comment this out, because I was getting "already in $digest" errors. 

This is the only thing left to solve in this PR.
Anybody feels like debugging this?

RawKStar77 replied:
I fetched the branch and can't get the digest error you talked about to happen, but you could maybe try:
`$timeout -> ctrl.$setViewValue inputEl.val()`

csillag replied:
Detailed instructions to reproduce:
 - Uncomment [this line]https://github.com/hypothesis/h/blob/export-showframe/h/static/scripts/directives/markdown.coffee#L339
 - Run code
 - Select something in the document
 - Press the adder button

Thanks.

csillag replied:
While debugging this *digest*ion problem, you might want to comment out [this line](https://github.com/hypothesis/h/blob/export-showframe/h/static/scripts/guest.coffee#L386), to reduce the number of things going on simultaneously in Angular.

gergely-ujvari replied:
I cannot reproduce this. I followed the instructions, no $digest error.

gergely-ujvari replied:
Aha... I see.
I get no $digest error in chrome, but I get the error in chromium.
We should consider an angularjs upgrade maybe?

csillag replied:
> 
> We should consider an angularjs upgrade maybe?
> 

That would be super sweet.

gergely-ujvari replied:
No, now I can see it in chrome but not always.. it seems the error only comes when you select a specific part. Interesting, for example, in our welcome page I select "started" to get it, but selecting "Getting" in chrome does not trigger it. (So the two browsers are behaving a bit different, but we can have the error in chrome too)

RawKStar77 replied:
Ah, I see it, thanks @gergely-ujvari for the pointers... that is weird. I don't claim to be an angular expert but rewriting:
```
    inputEl.bind 'blur change keyup', ->
      ctrl.$setViewValue inputEl.val()
      scope.$digest()  # TODO: this shouldn't be a problem
```
To be this:
```
    inputEl.bind 'blur change keyup', ->
      $timeout-> ctrl.$setViewValue inputEl.val()
```

Solved the problem, and everything still works even with Math rendering, so that could be a possible solution.

RawKStar77 replied:
It is really weird that it happens when you select some things but not others...

tilgovi commented:
Now the annotation ui sync service looks much better, don't you agree?

csillag commented:
> Now the annotation ui sync service looks much better, don't you agree?

I do. See also https://github.com/hypothesis/h/pull/2039 for making the bridge a factory (as suggested), so that I can inject in directly into the host service (as suggested)

csillag commented:
> The only thing that instantiates `Bridge` is the cross frame service. I don't see any reason why it can't receive a bridge instance and then the bridge itself can be constructed by an angular factory.

I did that in https://github.com/hypothesis/h/pull/2039. As soon as you merge it, I can rebase this on top of that, so that this will be done here, too.

csillag commented:
Rebased to master, which has the bridge changes. 
Updated injection scheme, as suggested earlier.

I think now this does what I wanted it to do, except
 
  * Tests need to be updated
  * There is the digest problem, which forced me to comment out a line in the markdown directive

csillag commented:
@tilgovi, are you satisfied with how this looks like now?

(I mean, except the digest cycle problem, and the broken tests.)

nickstenning commented on h/static/scripts/host-service.coffee:
```diff
@@ -0,0 +1,34 @@
+# @name host
+#
+# @description
+# The `host` service relay the instructions the sidebar needs to send
```
relays

nickstenning commented on h/static/scripts/host-service.coffee:
```diff
@@ -0,0 +1,34 @@
+#
+# @description
+# The `host` service relay the instructions the sidebar needs to send
+# to the host doument. (As opposed to all guests)
```
document

nickstenning commented on h/static/scripts/host-service.coffee:
```diff
@@ -0,0 +1,34 @@
+###
+class HostService
+
+  this.inject = [ '$window', 'bridge' ]
```
Can we cut down a bit on whitespace here? `this.inject = ['$window', 'bridge']`.

nickstenning commented:
HostService needs tests, as I'm sure you know.

csillag commented:
@nickstenning, thanks for catching the typos. Will fix those. About the tests: yes, I intend to fix/add them, as soon as we can settle the discussion on the code.

csillag commented:
To repeat, I'll add the tests for this when @tilgovi confirms that he likes the code changes in this PR.

tilgovi commented:
Go for it
On Mar 16, 2015 9:44 AM, "Kristof Csillag" <notifications@github.com> wrote:

> To repeat, I'll add the tests for this when @tilgovi
> <https://github.com/tilgovi> confirms that he likes the code changes in
> this PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hypothesis/h/pull/2038#issuecomment-81790301>.
>

csillag commented:
> HostService needs tests, as I'm sure you know.

Added tests.

tilgovi commented:
I think this looks good. Needs a rebase. If the digest issue doesn't start with this PR let's ignore and remove that comment, though.

csillag commented:
> If the digest issue doesn't start with this PR let's ignore and remove that comment, though.

It does not start with this PR, but this PR uncovers it. (It was not triggered before). I submitted a separate PR to remove the issue. (#2065) I'll wait for it to be merged, then I'll rebase this.

csillag commented:
Closing in favor of #2066.
